`pick-deep`: Fix the case when the key or element is an union type (#1224)

taiyakihitotsu opened the pull request:
This PR fixes #1224.  
(I’m opening this because I previously sent a message about the issue, but it seems that no one has started working on it.)

## Current Status
https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3389196769 -> https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3756828299 -> https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3764800851

## Approach

The issue can be broken down into two separate bugs:

 - `PickDeep` fails to pick values when the object’s properties are of a union type. (2025/09/29 ... ok)
 - `PickDeep` doesn't properly merge results when the path is a union type. ~~(2025/09/29 ... it has a [bug](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3341992076))~~ (2025/10/01 ... [ok](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3355791883))

The first was fixed by applying distribution over the union type in the object.

~~For the second issue, the fix involves evaluating each path in the union individually, obtaining the result as a union of objects, and then merging them using `MergeNarrow`.~~ (2025/10/01 ... [See](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3355791883))


## related issue

#1224, #1223, #880 ([See](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3756828299))

~~I'm not sure whether the expected value shown in #1223 is truly correct according to the `PickDeep` specification, so I’ve decided to leave that issue for now.~~ (2026/01/16 ... [See](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3756828299))

~~(In the current commit, the result for that case would be `{arr: []}`. I think this is caused by `MergeNarrow`.)~~ (2025/10/01 ... [See](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3355791883))

~~If I should include a fix for #1223 in this PR as well, please let me know.~~ (2026/01/16 ... [See](https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3756828299))



## note
I rewrote the tests to be more strict, using the following pattern:

```typescript
expectType<true>({} as IsEqual<Actual, Expected>)
```

This is because some previous tests using `expectType<T>(value)` were passing unintentionally.

taiyakihitotsu committed:
pick-deep: Fix the case when the key or element is an union type (#1224)

sindresorhus committed:
Update pick-deep.ts

taiyakihitotsu committed:
resolve conflict.

som-sm commented:
@taiyakihitotsu Thanks for the PR! I’m a bit occupied with festivities, I’ll try to review it sometime next week.

taiyakihitotsu committed:
Fix #1223

taiyakihitotsu commented:
I've added a fix commit for #1223.

@sindresorhus
Thanks!

@som-sm
Understood, I’ll wait. Enjoy!

taiyakihitotsu commented:
I found a bug in this case and will fix it as well.

```typescript
type unionKeyUnionObject_Actual = PickDeep<{obj: {a: string; b: number; c: boolean} | {d: 0}}, `obj.${'b' | 'c' | 'd'}`>; // {obj: {b: number; d: 0; c: boolean}}
```

taiyakihitotsu commented:
And `LastOfUnion` is also defined in

https://github.com/sindresorhus/type-fest/blob/ee29ef7b9d6280126562baae95795d4ed97ff3c8/source/union-to-tuple.d.ts#L13

I'll move it to internal if ok to combine them into one PR.

taiyakihitotsu commented:
I’ve got an idea for the fix, but not sure when I can get back to it, so I leave this as a draft for now.
Maybe I’ll pick it up again in a week or so.

taiyakihitotsu committed:
Fix: union type including same shape pbjects

som-sm commented:
> The current `IsNever` implementation has a bug:

Is there a simpler reproduction of where `IsNever` fails?

taiyakihitotsu commented:
@som-sm

It may only fail when using recursion with `LastOfUnion<T>` and `IsNever`. This is the same issue we’re currently facing with `IsEqual`.

## **[Current `IsEqual`](https://github.com/sindresorhus/type-fest/commit/785549f36465e3f3d99a08832784b603261f74f2)**
### Fixed
 - This fixed the [three cases we discussed](https://github.com/sindresorhus/type-fest/pull/1231#issuecomment-3313480326)
 - `never` and `unknown` pattern fixed. https://github.com/sindresorhus/type-fest/pull/1251
### Failed
 - `never` returned `LastOfUnion`.

## Internal `_IsEqual`
```typescript
type _IsEqual<A, B> =
	(<G>() => G extends A & G | G ? 1 : 2) extends
	(<G>() => G extends B & G | G ? 1 : 2)
		? true
		: false;
 ```   
This has bugs above the `fixed` part.
But it can detect well of `never` of `LastOfUnion` if defining like:

```typescript
export type CurrentIsNever<T> = _IsEqual<T, never>
```

### whole code
I'm researching it in `source/is-equal.d.ts`.

```typescript
export type IsEqual<A, B> =
	[A, B] extends [infer AA, infer BB]
		? [AA] extends [never]
			? [BB] extends [never]
				? true
				: false
			: [BB] extends [never]
				? false
				: _IsEqual<AA, BB>
		: false;

import type {UnionToIntersection} from './union-to-intersection.d.ts'
type LastOfUnion<T> = UnionToIntersection<T extends any ? () => T : never> extends () => (infer R)	? R : never;

export type PreviousIsNever<T> = [T] extends [never] ? true : false;
export type IsNever<T> = IsEqual<T, never>
export type CurrentIsNever<T> = _IsEqual<T, never>

type PreviousUnionRecursion<T, R = {}> = LastOfUnion<T> extends infer L ? PreviousIsNever<L> extends false ? PreviousUnionRecursion<Exclude<T, L>, R | L> : R : never;
type UnionRecursion<T, R = {}> = LastOfUnion<T> extends infer L ? IsNever<L> extends false ? UnionRecursion<Exclude<T, L>, R | L> : R : never;
type CurrentUnionRecursion<T, R = {}> = LastOfUnion<T> extends infer L ? CurrentIsNever<L> extends false ? CurrentUnionRecursion<Exclude<T, L>, R | L> : R : never;

// @ts-expect-error
type PrevTest = PreviousUnionRecursion<string | number>;
// Type instantiation is excessively deep and possibly infinite.

// @ts-expect-error
type FixedTest = UnionRecursion<string | number>; 
// Type instantiation is excessively deep and possibly infinite.

type CurrentFixedTest = CurrentUnionRecursion<string | number>; // string | number | {}

// This version fails the `equalWrappedTupleIntersectionToBeNeverAndNeverExpanded` test in `test-d/is-equal.ts`.
type _IsEqual<A, B> =
	(<G>() => G extends A & G | G ? 1 : 2) extends
	(<G>() => G extends B & G | G ? 1 : 2)
		? true
		: false;
```

There are two plans. One is to redefine `IsNever` as shown below in `CurrentIsNever`, and at the same time make `_IsEqual` internal like `InternalIsEqual`. This solution has already been discovered, is simple to implement, and maintains consistency in terms of `never` checking. However, it sacrifices some theoretical rigor.

The other plan is to fix `IsEqual` and then redefine `IsNever` using the revised `IsEqual`. So far, this approach has not been successful.

taiyakihitotsu commented:
@som-sm

```typescript
type LastOfUnion<T> =
  UnionToIntersection<
    T extends any
      ? () => any
      : never> extends () => (infer R)
    ? R
    : never;

type a = LastOfUnion<never>; // unknown
```

OK I see that this returns `unknown`, and `_IsEqual<never, unknown>` returns `true` unintentionally.
So `LastOfUnion<U> extends infer L` and `L extends Something` works as expected, `IsNever` using `_IsEqual` works fine.
But the current `IsEqual` returns `false` for `L` even if it's expected to be `never` (`L` is `unknown` in this case though).

The current (at [590efae8223ed1ac61855840dff114360b7c6807](https://github.com/sindresorhus/type-fest/commit/590efae8223ed1ac61855840dff114360b7c6807)) definition of `IsEqual` and `IsNever` aren't incorrect.
It's not a bug in `IsNever`.

Apologies for my misunderstanding.

In any case, `LastOfUnion` is already used in [source/union-to-tuple.d.ts](https://github.com/sindresorhus/type-fest/blob/590efae8223ed1ac61855840dff114360b7c6807/source/union-to-tuple.d.ts#L13), and this commit defines it redundantly.
So I want to move it to internal.

```typescript
type test = LastOfUnion<Exclude<string, string>> extends infer U ? PreviousIsNever<U> : 'false';
// ^ false

export type PreviousIsNever<T> = [T] extends [never] ? true : false;
```

taiyakihitotsu committed:
Fix: internal LastOfUnion

taiyakihitotsu committed:
refactor: LastOfUnion moved to internal

taiyakihitotsu committed:
refactor: Using IsNever instead of IsEqual<T, never>

sindresorhus committed:
Update pick-deep.d.ts

taiyakihitotsu committed:
chore: sync with main

taiyakihitotsu committed:
fix: resolve merge conflict

taiyakihitotsu committed:
Merge branch 'main' into fix/pick-deep

taiyakihitotsu committed:
refactor: use internal StringToNumber

taiyakihitotsu commented:
@som-sm

Sorry for leaving this hanging.

 - Merged the latest main branch.
 - Switched to the internal `StringToNumber`.

The functionality hasn’t changed since https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3389196769.

---

I just wanted to share the update.
Feel free to take a look whenever you have a moment.

sindresorhus commented on source/pick-deep.d.ts:
```diff
@@ -77,76 +226,109 @@ type Street = PickDeep<Configuration, 'userConfig.address.1.street2'>;
+	NextParent extends infer NextParentArray extends UnknownArray
+		/* NextParent: array */
+		? IsEqual<IsTuple<NextParentArray>, false> extends true
+			? NextParentArray extends Array<infer _>
```
```suggestion
			? NextParentArray extends readonly unknown[]
```

taiyakihitotsu committed:
fix: keep readonly properties.

taiyakihitotsu committed:
refactor: use local infer type (SubTree)

taiyakihitotsu committed:
test: check if null and undefined are preserved.

taiyakihitotsu commented:
@sindresorhus

Thanks for your review!
I apologize for missing some cases regarding `readonly` properties and leaving trivial mistakes in the code.
I’ve addressed all of your suggestions.

---

### my note

This test case which you suggested was not passed.

```typescript
// Test: readonly array should preserve readonly modifier when picking into elements
type ReadonlyArrayContainer = {
	items: ReadonlyArray<{a: 1; b: 2}>;
};
type readonlyArrayPick_Actual = PickDeep<ReadonlyArrayContainer, `items.${number}.a`>;
type readonlyArrayPick_Expected = {items: ReadonlyArray<{a: 1}>};
expectType<true>({} as IsEqual<readonlyArrayPick_Actual, readonlyArrayPick_Expected>);
```

This review's code (https://github.com/sindresorhus/type-fest/pull/1241#discussion_r2695598791) fixed this tests (https://github.com/sindresorhus/type-fest/pull/1241#discussion_r2695593864):

---

This PR closes #1224 (and potentially #1223).
It may also potentially close #880, but there is an ongoing discussion that has been suspended (https://github.com/sindresorhus/type-fest/issues/880#issuecomment-2121625777).

To clarify this point, I’ve added a comment and a test case (in this commit) to ensure that `null` and `undefined` are preserved for this scenario, as previously discussed here: https://github.com/sindresorhus/type-fest/issues/880#issuecomment-2115864014.
At least for this use case, it seems to work without issues.

taiyakihitotsu committed:
fix: ensure `IsKeyOf<Collection, never>` returns `false` properly.

taiyakihitotsu committed:
docs: add use cases

taiyakihitotsu commented:
I've fixed `IsKeyOf` because `IsKeyOf<Coll, never>` returned `never` incorrectly.
It causes a problem that `IsKeyOf<Coll, never>` could satisfy `extends true` unintentionally.

The other status has not been changed from https://github.com/sindresorhus/type-fest/pull/1241#issuecomment-3756828299.

taiyakihitotsu commented:
Close this because there are too many conflicts.
I'll open a new PR with this patch.
